include/sparrow/buffer_adaptor.hpp

```diff
@@ -0,0 +1,499 @@
+    {
+        return iterator(data() + size());
+    }
+
+    template <typename T, typename U>
+        requires(sizeof(U) < sizeof(T))
+    constexpr typename buffer_adaptor<T, U>::const_iterator buffer_adaptor<T, U>::begin() const noexcept
+    {
+        return const_iterator(data());
+    }
+
+    template <typename T, typename U>
+        requires(sizeof(U) < sizeof(T))
+    constexpr typename buffer_adaptor<T, U>::const_iterator buffer_adaptor<T, U>::end() const noexcept
+    {
+        return const_iterator(data() + size());
+    }
+
+    template <typename T, typename U>
+        requires(sizeof(U) < sizeof(T))
+    constexpr typename buffer_adaptor<T, U>::const_iterator buffer_adaptor<T, U>::cbegin() const noexcept
+    {
+        return begin();
+    }
+
+    template <typename T, typename U>
+        requires(sizeof(U) < sizeof(T))
+    constexpr typename buffer_adaptor<T, U>::const_iterator buffer_adaptor<T, U>::cend() const noexcept
+    {
+        return end();
+    }
+
+    template <typename T, typename U>
+        requires(sizeof(U) < sizeof(T))
+    constexpr typename buffer_adaptor<T, U>::reverse_iterator buffer_adaptor<T, U>::rbegin() noexcept
+    {
+        return reverse_iterator(end());
+    }
+
+    template <typename T, typename U>
+        requires(sizeof(U) < sizeof(T))
+    constexpr typename buffer_adaptor<T, U>::reverse_iterator buffer_adaptor<T, U>::rend() noexcept
+    {
+        return reverse_iterator(begin());
+    }
+
+    template <typename T, typename U>
+        requires(sizeof(U) < sizeof(T))
+    constexpr typename buffer_adaptor<T, U>::const_reverse_iterator buffer_adaptor<T, U>::rbegin() const noexcept
+    {
+        return const_reverse_iterator(end());
+    }
+
+    template <typename T, typename U>
+        requires(sizeof(U) < sizeof(T))
+    constexpr typename buffer_adaptor<T, U>::const_reverse_iterator buffer_adaptor<T, U>::rend() const noexcept
+    {
+        return const_reverse_iterator(begin());
+    }
+
+    template <typename T, typename U>
+        requires(sizeof(U) < sizeof(T))
+    constexpr typename buffer_adaptor<T, U>::const_reverse_iterator
+    buffer_adaptor<T, U>::crbegin() const noexcept
+    {
+        return rbegin();
+    }
+
+    template <typename T, typename U>
+        requires(sizeof(U) < sizeof(T))
+    constexpr typename buffer_adaptor<T, U>::const_reverse_iterator buffer_adaptor<T, U>::crend() const noexcept
+    {
+        return rend();
+    }
+
+    // Capacity
+
+    template <typename T, typename U>
+        requires(sizeof(U) < sizeof(T))
+    constexpr buffer_adaptor<T, U>::size_type buffer_adaptor<T, U>::size() const noexcept
+    {
+        const double new_size = static_cast<double>(_buffer.size()) * _sizeof_ratio;
+        SPARROW_ASSERT(
+            std::trunc(new_size) == new_size,
+            "The size of the buffer is not a multiple of the size of the new type"
+        );
```

Comment:
Should we express this as a requirement on the size of `T` and `U` instead? I don't see a case where we would convert a type into another with a non integer size of ratio.

Reply:
Indeed, my assumption was that something stored as uint8_t and that we cast in another type would actually be stored as uint8_t for technical reasons (dynamic type, deserialization), and that such a case would never happen. But that mmight not be the case, and anyway, we want to catch this kind of bugs.